Require undef in BlockSparseArray constructors

mtfishman opened the pull request:
Fixes #69. Related to https://github.com/ITensor/SparseArraysBase.jl/pull/33.

This also introduces `sparsemortar`, which is like [`BlockArrays.mortar`](https://juliaarrays.github.io/BlockArrays.jl/stable/lib/public/#BlockArrays.mortar) but for constructing a `BlockSparseArray`.

mtfishman commented:
@lkdvos I'd be curious to get your feedback on this, particularly the name `sparsemortar`.

lkdvos commented on src/blocksparsearray/blocksparsearray.jl:
I think I would maybe consider altering the constructor call chain a bit, more similar to how it is in SparseArraysBase:

**Inner constructors**:
- a bare constructor which is fully type-specified, with `undef`. This generates the containers, and gives you a fresh empty blocksparsearray
- an (unchecked) constructor to wrap data

**Outer constructors**:
- chain of `undef` functions that simply add the correct type parameters and pass everything on to the inner `undef` constructor.
- `BlockSparseArray(data, ...)` that determines the type parameters from the data and pass that to the inner `data` constructor.
- (possibly) `BlockSparseArray{...}(data, ...)` that converts the data to the correct types to be compatible with the requested types.

mtfishman replied:
Basically, I think the core of the issue is that `BlockSparseArray(a::AbstractArray, axes)` could have two meanings, either:
1. `a` is an array we want to convert to a `BlockSparseArray`.
2. `a` is an array of blocks that we want to "mortar" into a `BlockSparseArray`.

I'm trying to stick to it meaning 1. (which is how BlockArrays.jl is designed).

mtfishman replied:
> I think I tend to think of "constructors" as being type 1 or type 2 from your list, and maybe consider orthogonalizing that to type 3 and 4 a bit, since otherwise we have to cram quite a lot of functionality into one function.

Agreed. For now I've made 1. into a private constructor `_BlockSparseArray`.

> I'd say 3 is a harder question, but I would also say technically it's not really a conversion, since it is not bijective: if you need the `ax` argument, this is extra information. This seems to indicate that it is more than a constructor/converter, and perhaps means it deserves it's own function? On the other hand, I do agree that it really is _almost_ a conversion, so we could also decide to treat it that way...

That's true, that one is a little bit on the line, here I'm breaking the tie by going with the convention that BlockArrays.jl uses. Maybe for that one it is better characterized as a wrapping or reinterpretation rather than strictly as a conversion. And for BlockSparseArrays it is a bit more subtle since we also want to drop zero blocks. But I think mostly the point is to try to decrease the number of different concepts or types of conversions/constructions we try to fit into the `BlockSparseArray` constructor itself, this one seems pretty reasonable, though probably it would be nice to have a function `blocksparse` that covers that functionality but can be extended to other types (and also other types of inputs, like lists of nonzero blocks).

lkdvos commented on src/blocksparsearray/blocksparsearray.jl:
```diff
@@ -27,142 +24,140 @@ const BlockSparseVector{T,A<:AbstractVector{T},Blocks<:AbstractVector{A},Axes<:T
 
-function BlockSparseArray(
+"""
+    sparsemortar(blocks::AbstractArray{<:AbstractArray{T,N},N}, axes) -> ::BlockSparseArray{T,N}
```
Some musings, even though I think having this function probably also makes sense:

- Can we get away with overloading `mortar` in the case of sparse arrays of blocks, or do we think the piracy in `mortar(::Dictionary)` is a bit much. From the docstring of `mortar`, it says that it is an "inverse" of `blocks`, so that would keep that property.
- Do we want to allow the `sparsemortar(::Dictionary)` call? It might be fair to require a manual `SparseArrayDOK` call beforehand.
- Can we define `sparse(::BlockArray)` such that `sparsemortar(args...) = sparse(mortar(args...))`? This is related to the question of simply overloading `mortar` for sparse arrays directly, since I'm guessing this indirection probably does not hurt performance all that much.

mtfishman replied:
>     * Can we get away with overloading `mortar` in the case of sparse arrays of blocks, or do we think the piracy in `mortar(::Dictionary)` is a bit much. From the docstring of `mortar`, it says that it is an "inverse" of `blocks`, so that would keep that property.

I considered that, but I would be worried about that kind of piracy, why don't we hold off on that for now.

>     * Do we want to allow the `sparsemortar(::Dictionary)` call? It might be fair to require a manual `SparseArrayDOK` call beforehand.

That I wasn't sure about. I think it is convenient to be able to specify the list of blocks directly rather than having to construct the `SparseArrayDOK`. Maybe we should give it a different name, what about `blocksparse(::Dictionary, axes)`? There's some precedence for that with [`SparseArrays.sparsevec`](https://sparsearrays.juliasparse.org/dev/#SparseArrays.sparsevec), and also it has the same spirit as the [`SparseArrays.sparse`](https://sparsearrays.juliasparse.org/dev/#SparseArrays.sparse) constructor that takes lists of nonzero element locations and values.

>     * Can we define `sparse(::BlockArray)` such that `sparsemortar(args...) = sparse(mortar(args...))`? This is related to the question of simply overloading `mortar` for sparse arrays directly, since I'm guessing this indirection probably does not hurt performance all that much.

That's an interesting point, I hadn't considered that. One nitpick I would have with that proposal is using `sparse(...)` for that functionality. I would expect `sparse(::BlockArray)` would either output a `SparseArrayDOK` or a `BlockSparseArray` where the blocks are `SparseArrayDOK` (i.e. it would sparsify the block structure _and_ the blocks), maybe `blocksparse(::BlockArray)` would be a clearer name which would just sparsify the block structure and preserve the types of the nonzero blocks. So then maybe we could have `sparsemortar(args...) = blocksparse(mortar(args...))`. I was considering calling it `blocksparsemortar`, but I figured that the name `mortar` already implies it is blocked, so the term "sparse" in the name implicitly refers to the sparsity of the block pattern, not the elements.

mtfishman replied:
Regarding the first point, I think I would feel comfortable defining `mortar(::SparseArrayDOK)` so that it returns `BlockSparseArray`, but maybe we can keep `sparsemortar` for types we don't own, for example `sparsemortar(::Diagonal)` could make a `BlockSparseArray` while `mortar(::Diagonal)` would make a `BlockArray`.

mtfishman replied:
I basically agree with you, and think we should almost always use `BlockSparseArray{Float64}(undef, axes)` and then fill with blocks, or construct a `SparseArrayDOK`. I definitely agree that the fact that it stores a `Dictionary` inside should be an implementation detail, but that can be done in multiple ways in practice.

I still think it would be nice to be able to write something like `blocksparse(Dict([Block(1, 1) => randn(2, 2), Block(2, 2) => randn(2, 2)]), [2, 2], [2, 2])`, but we could hold off on that for now. I would however like to anticipate that we _might_ want that, and avoid defining a constructor now that would make it awkward to define that later.

I think at this point I'll take another stab at this PR and start by being pretty conservative about the constructors I define, taking these various points into account.

It does make we wonder if we should also be stricter about `SparseArrayDOK` constructor, and maybe the constructor from `Dictionary`/`Dict` should be called `sparse`, and `SparseArrayDOK` only has an `undef` constructor or acts as a converter from other array types (i.e. `SparseArrayDOK(a::AbstractArray)` returns a sparse version of `a`).

lkdvos replied:
I do like this last point. In general, I tend to feel like often there is just too much functionality crammed into constructors, which would be better if it is kept more separate.

mtfishman replied:
> I agree with this last point. In general, I tend to feel like often there is just too much functionality crammed into constructors, which would be better if it is kept more separate.

Alright, let's plan on changing that in SparseArraysBase.jl, I'll make an issue about it.

mtfishman replied:
I raised an issue here: https://github.com/ITensor/SparseArraysBase.jl/issues/42.

lkdvos reviewed (commented):
Overall definitely looks good to me.

I left some comments about the organization of the constructor call chain, but keep in mind that I'm not sure if that structure is *standard practice* or simply something I have picked up over the years which I have found to be quite robust and stable.
Bottom line, this is a bit subjective anyways, but I do like the clear separation of responsibilities.

Considering the `sparsemortar` name, I think that's good, and in line with `sparserand` and `sparsezeros` so that seems consistent.
I am wondering a bit if we can get away with not having that function at all, and repurposing `sparse(mortar(args...))` instead, which is only a few characters longer and reuses functionality we would probably have to have anyways. (in which case we can still define it ofcourse). I don't think it matters too much of course, just wanted to share some thoughts 😉

mtfishman commented on src/blocksparsearray/blocksparsearray.jl:
```diff
@@ -27,142 +24,140 @@ const BlockSparseVector{T,A<:AbstractVector{T},Blocks<:AbstractVector{A},Axes<:T
 ) where {T,N,A<:AbstractArray{T,N}}
   blocks = default_blocks(A, axes)
-  return BlockSparseArray{T,N,A}(blocks, axes)
+  return sparsemortar(blocks, axes)
```
That's a good point, I'll change that. One reason I avoided that was because I wanted to create a clear distinction between which constructors accept blocks vs. build the block sparse array from scratch. Maybe what I'll do here is make a private "inner constructor" called `_BlockSparseArray` (like [`BlockArrays._BlockArray`](https://github.com/JuliaArrays/BlockArrays.jl/blob/v1.4.0/src/blockarray.jl#L65-L69)) to make it clear this is calling an inner constructor.

mtfishman commented:
@lkdvos in the latest I:
1. introduced `_BlockSparseArray` as an inner constructor,
2. deleted versions of `BlockSparseArray` that accept blocks,
3. deleted `sparsemortar(::Dictionary, axes)` and also the version that accepted lists of block locations and data, so now there is only `sparsemortar(blocks::AbstractArray{<:AbstractArray}, axes)`, and
4. introduced `SparseArrayDOK{Matrix{Float64}}(undef_blocks, [2, 3], [2, 3])` as a helper function for constructing block sparse arrays.

lkdvos commented on src/blocksparsearray/blockdiagonalarray.jl:
```diff
@@ -12,9 +12,7 @@ const BlockSparseDiagonal{T,A<:AbstractBlockSparseVector{T}} = Diagonal{T,A}
-  return BlockSparseArray(
-    Diagonal(blocks), (blockedrange(size.(blocks, 1)), blockedrange(size.(blocks, 2)))
-  )
+  return sparsemortar(Diagonal(blocks), size.(blocks, 1), size.(blocks, 2))
```
I think this one is also secretly a constructor (`BlockSparseMatrix{T,A,Diagonal{A,V},Ax} where {T,A,V<:AbstractVector{A},Ax}`), so maybe this should also use the inner constructor?

lkdvos commented on src/blocksparsearray/blocksparsearray.jl:
```diff
@@ -1,168 +1,223 @@
-# TODO: Delete this.
-## using BlockArrays: blocks
+"""
+    SparseArrayDOK{T}(::UndefBlocksInitializer, axes)
+    SparseArrayDOK{T,N}(::UndefBlocksInitializer, axes)
```
```suggestion
    SparseArrayDOK{T}(undef_blocks, axes)
    SparseArrayDOK{T,N}(undef_blocks, axes)
```

I think this is a bit more helpful, and typically also how `undef` is mentioned

lkdvos replied:
If necessary, we could also consider adding
```
See also: [`undef_blocks`](@ref BlockArrays.undef_blocks)`
```

mtfishman replied:
Good suggestion, I'll change that.

mtfishman replied:
It looks like you need some external tools to cross reference the docs of a function defined in another package so I just made a link manually for now.

lkdvos commented on src/blocksparsearray/blocksparsearray.jl:
```diff
@@ -1,168 +1,223 @@
+) where {T,N}
+  return SparseArrayDOK{T,N}(undef_blocks, blockedrange.(dims))
+end
+
```
This probably has the usual `Vararg{X,N}` ambiguity/unbound parameter issues when `N=0`. Something I've started doing to combat this is things like:
```julia
function f(::UndefBlocksInitializer, dim::I, dims::I...) where {I<:...}
```
which effectively changes this to `Vararg{I,N}` with `N >= 1`

mtfishman replied:
That's a good point, I'll change that.

lkdvos commented on src/blocksparsearray/blocksparsearray.jl:
```diff
@@ -1,168 +1,223 @@
   A<:AbstractArray{T,N},
   Blocks<:AbstractArray{A,N},
-  Axes<:Tuple{Vararg{AbstractUnitRange,N}},
+  Axes<:Tuple{Vararg{AbstractUnitRange{<:Integer},N}},
```
Is this a change we want? Should it have a separate PR? It seems like it might affect design choices for symmetric block arrays at some point.

mtfishman replied:
I think those axes will still have integer element types.

lkdvos replied:
I'm definitely okay with it, just double-checking since this seems like a consequential decision. (Which can easily be undone if ever necessary)

lkdvos commented on src/blocksparsearrayinterface/map.jl:
```diff
@@ -94,23 +96,17 @@ function map_zero_dim! end
+  blocks_a = blocks(a)
+  stored_indices = collect(eachstoredindex(blocks_a))
+  stored_blocks = map(I -> f(blocks_a[I]), stored_indices)
+  blocks_a′ = SparseArrayDOK{eltype(stored_blocks)}(undef_blocks, axes(a))
```
Can we immediately construct a `BlockSparseArray` instead?

mtfishman replied:
That's how I had written it, partially this rewrite was to try out this new `SparseArrayDOK` constructor, but probably writing it in terms of `BlockSparseArray` is better, I'll change it back.

mtfishman commented:
@lkdvos I've update the PR based on your last round of comments, this is good to go from my end. Let me know if you have any more comments.

lkdvos commented on src/blocksparsearray/blocksparsearray.jl:
```diff
@@ -1,168 +1,225 @@
+blocked axes `axes`.
+
+Note that `undef_blocks` is defined in
+[BlockArrays.jl](https://juliaarrays.github.io/BlockArrays.jl/stable/lib/public/#BlockArrays.undef_blocks)
```
We could consider re-exporting that too if it turns out we are using this often

mtfishman replied:
Right, that touches on a broader question I've been grappling with about how much we want to re-export things. So far my strategy with these new packages is to be very conservative about what we export, including not re-exporting anything and expecting users to load the packages where the objects originated themselves, but once the dust settles a bit and we start using these packages more I'm open to some targeted re-exports.
